### PR TITLE
EOS-23641: Add CRC support for btree.

### PR DESCRIPTION
**Pull request by NikhilBirgade**
Signed-off-by: Nikhil Kumar Birgade <nikhil.birgade@seagate.com>

### CODE REVIEW

File: btree/btree.c [hunk 1]
```diff
@@ -3108,6 +3108,130 @@ static void ff_rec_del_credit(const struct nd *node, m0_bcount_t ksize,
...
+ * A btree node consists of node header, keys, values and dictionary(in case
+ * of variable key/value size format). Due to bugs in the code, storage bitrot,
+ * memory bitrot etc., the btree node can get corrupted. CRC will be used to
+ * detect such corruption in the btree node.
+ *
+ * Common Assumptions:
+ *	a. For internal nodes, Motr will calculate CRCs of only Keys.
+ *	b. For internal nodes, Motr will not calculate CRCs of values instead
+ *	some sanity checks will be used to verify the values.
+ *	c. CRC is updated when a record us added or deleted.
```
Comment (ydb242) [inline btree/btree.c:3123]: s/us/is

File: btree/btree.c [hunk 2]
```diff
@@ -3108,6 +3108,130 @@ static void ff_rec_del_credit(const struct nd *node, m0_bcount_t ksize,
...
+ *	d. While traversing down the tree for the first time  for any operation
```
Comment (ydb242) [inline btree/btree.c:3124]: extra space after time

File: btree/btree.c [hunk 3]
```diff
@@ -3108,6 +3108,130 @@ static void ff_rec_del_credit(const struct nd *node, m0_bcount_t ksize,
...
+ *	(get/delete/put), CRC will not be checked.
+ *	e. For get and delete operation if tree traversal reaches leaf node and
```
Comment (ydb242) [inline btree/btree.c:3126]: comma after operation

File: btree/btree.c [hunk 4]
```diff
@@ -3108,6 +3108,130 @@ static void ff_rec_del_credit(const struct nd *node, m0_bcount_t ksize,
...
+ *	do not find the record then tree traversal will be re-initiated with CRC
```
Comment (ydb242) [inline btree/btree.c:3127]: s/do/does

File: btree/btree.c [hunk 5]
```diff
@@ -3108,6 +3108,130 @@ static void ff_rec_del_credit(const struct nd *node, m0_bcount_t ksize,
...
+ * |Node Header| K0 | K1 | K2 |  ----->                <----- | V2 | V1 | V0 |
+ * |           |    |    |    |                               |    |    |    |
+ * |           |    |    |    |                               |    |    |    |
+ * +-----------+----+----+----+-------------------------------+----+----+----+
+ *
+ *	b. Motr will calculate individual checksum for each key and save it next
+ *	to the key in the node.
+ *		i.Checksum of all the keys are verified when node is loaded from
+ *		the disk.
+ *		ii.Only the checksum of the newly added key is calculated and
```
Comment (ydb242) [inline btree/btree.c:3174]: space after ii.

File: btree/btree.c [hunk 6]
```diff
@@ -3108,6 +3108,130 @@ static void ff_rec_del_credit(const struct nd *node, m0_bcount_t ksize,
...
+ *	f. For insert operation, any corruption in node can result in tree
+ *	traversal pointing to wrong memory location. Since CRC is not verified
+ *	during tree traversal, we cannot determine if there was any corruption
+ *	and the memory location(at the end of tree traversal) is wrong. We have
+ *	two options to handle this issue:
+ *		i. We do not worry about this issue and add the new record at
+ *		whatever location we have found and continue. This is because
+ *		the path is corrupted anyways and there is no way to recover the
+ *		bad path and reach the correct location.
+ *		ii.We worry about this issue if we have a way to fix the tree
```
Comment (ydb242) [inline btree/btree.c:3138]: space after ii.

File: btree/btree.c [hunk 7]
```diff
@@ -3108,6 +3108,130 @@ static void ff_rec_del_credit(const struct nd *node, m0_bcount_t ksize,
...
+ *
+ * +------+----+----+----+----+-------------------------------+----+----+----+
+ * |      |CRC |    |    |    |                               |    |    |    |
+ * |      +----+    |    |    |                               |    |    |    |
+ * |Node Header| K0 | K1 | K2 |  ----->                <----- | V2 | V1 | V0 |
+ * |           |    |    |    |                               |    |    |    |
+ * |           |    |    |    |                               |    |    |    |
+ * +-----------+----+----+----+-------------------------------+----+----+----+
+ *
+ *	b. Mote will calculate individual checksum for each key and save it next
```
Comment (ydb242) [inline btree/btree.c:3212]: s/Mote/Motr

File: btree/btree.c [hunk 8]
```diff
@@ -3108,6 +3108,130 @@ static void ff_rec_del_credit(const struct nd *node, m0_bcount_t ksize,
...
+ *	to the key in the node.
+ *		i.Checksum of all the keys are verified when node is loaded from
+ *		the disk.
+ *		ii.Only the checksum of the newly added key is calculated and
```
Comment (ydb242) [inline btree/btree.c:3216]: space after ii.

File: btree/btree.c [hunk 9]
```diff
@@ -3108,6 +3108,130 @@ static void ff_rec_del_credit(const struct nd *node, m0_bcount_t ksize,
...
+ * |      |CRC |    |    |    |                               |    |    |    |
+ * |      +----+    |    |    |                               |    |    |    |
+ * |Node Header| K0 | K1 | K2 |  ----->                <----- | V2 | V1 | V0 |
+ * |           |    |    |    |                               |    |    |    |
+ * |           |    |    |    |                               |    |    |    |
+ * +-----------+----+----+----+-------------------------------+----+----+----+
+ *
+ *	b. Mote will calculate individual checksum for each key and save it next
+ *	to the key in the node.
+ *		i.Checksum of all the keys are verified when node is loaded from
```
Comment (ydb242) [inline btree/btree.c:3214]: space after i.

File: btree/btree.c [hunk 10]
```diff
@@ -3108,6 +3108,130 @@ static void ff_rec_del_credit(const struct nd *node, m0_bcount_t ksize,
...
+ * |      |CRC |    |    |    |                               |    |    |    |
+ * |      +----+    |    |    |                               |    |    |    |
+ * |Node Header| K0 | K1 | K2 |  ----->                <----- | V2 | V1 | V0 |
+ * |           |    |    |    |                               |    |    |    |
+ * |           |    |    |    |                               |    |    |    |
+ * +-----------+----+----+----+-------------------------------+----+----+----+
+ *
+ *	b. Motr will calculate individual checksum for each key and save it next
+ *	to the key in the node.
+ *		i.Checksum of all the keys are verified when node is loaded from
```
Comment (ydb242) [inline btree/btree.c:3172]: space after i.

File: btree/btree.c [hunk 11]
```diff
@@ -3108,6 +3108,130 @@ static void ff_rec_del_credit(const struct nd *node, m0_bcount_t ksize,
...
+/**
+ * CRC Support:
+ *
+ * A btree node consists of node header, keys, values and dictionary(in case
+ * of variable key/value size format). Due to bugs in the code, storage bitrot,
+ * memory bitrot etc., the btree node can get corrupted. CRC will be used to
+ * detect such corruption in the btree node.
+ *
+ * Common Assumptions:
+ *	a. For internal nodes, Motr will calculate CRCs of only Keys.
```
Comment (ydb242) [inline btree/btree.c:3120]: s/Keys/keys

Comment (NikhilBirgade) [reply btree/btree.c:3120]: Done.

File: btree/btree.c (hunk 1 above)
Comment (NikhilBirgade) [reply btree/btree.c:3123]: Done.

File: btree/btree.c (hunk 2 above)
Comment (NikhilBirgade) [reply btree/btree.c:3124]: Done.

File: btree/btree.c (hunk 3 above)
Comment (NikhilBirgade) [reply btree/btree.c:3126]: Done.

File: btree/btree.c (hunk 4 above)
Comment (NikhilBirgade) [reply btree/btree.c:3127]: Done.

File: btree/btree.c (hunk 6 above)
Comment (NikhilBirgade) [reply btree/btree.c:3138]: Done.

File: btree/btree.c (hunk 10 above)
Comment (NikhilBirgade) [reply btree/btree.c:3172]: Done.

File: btree/btree.c (hunk 5 above)
Comment (NikhilBirgade) [reply btree/btree.c:3174]: Done.

File: btree/btree.c (hunk 7 above)
Comment (NikhilBirgade) [reply btree/btree.c:3212]: Done.

File: btree/btree.c (hunk 9 above)
Comment (NikhilBirgade) [reply btree/btree.c:3214]: Done.

File: btree/btree.c (hunk 8 above)
Comment (NikhilBirgade) [reply btree/btree.c:3216]: Done.

**Review by ydb242 — APPROVED** [review]

**Review by RadhaGulhane13 — APPROVED** [review]

**Review by shashank-parulekar — CHANGES_REQUESTED** [review]

File: btree/btree.c [hunk 12]
```diff
@@ -3108,6 +3108,145 @@ static void ff_rec_del_credit(const struct nd *node, m0_bcount_t ksize,
...
+ *
+ * A btree node consists of node header, keys, values and dictionary(in case
+ * of variable key/value size format). Due to bugs in the code, storage bitrot,
+ * memory bitrot etc., the btree node can get corrupted. CRC will be used to
+ * detect such corruption in the btree node.
+ *
+ * Common Assumptions:
+ *	a. For internal nodes, Motr will calculate CRCs of only keys.
+ *	b. For internal nodes, Motr will not calculate CRCs of values instead
+ *	some sanity checks will be used to verify the values.
```
[the remark this answers is not in the document]
Comment (NikhilBirgade) [reply btree/btree.c:3122]: Done.

File: btree/btree.c [hunk 13]
```diff
@@ -3108,6 +3108,145 @@ static void ff_rec_del_credit(const struct nd *node, m0_bcount_t ksize,
...
+ *	e. For get and delete operation, if tree traversal reaches leaf node and
+ *	does not find the record then tree traversal will be re-initiated with
+ *	CRC check on.
+ *	f. For insert operation, any corruption in node can result in tree
+ *	traversal pointing to wrong memory location. Since CRC is not verified
+ *	during tree traversal, we cannot determine if there was any corruption
+ *	and the memory location(at the end of tree traversal) is wrong. We have
+ *	two options to handle this issue:
+ *		i. We do not worry about this issue and add the new record at
+ *		whatever location we have found and continue. This is because
```
[the remark this answers is not in the document]
Comment (NikhilBirgade) [reply btree/btree.c:3135]: Done.

File: btree/btree.c [hunk 14]
```diff
@@ -3108,6 +3108,145 @@ static void ff_rec_del_credit(const struct nd *node, m0_bcount_t ksize,
...
+ *		the path is corrupted anyways and there is no way to recover the
```
[the remark this answers is not in the document]
Comment (NikhilBirgade) [reply btree/btree.c:3136]: Done.

File: btree/btree.c [hunk 15]
```diff
@@ -3108,6 +3108,145 @@ static void ff_rec_del_credit(const struct nd *node, m0_bcount_t ksize,
...
+ *		bad path and reach the correct location.
```
[the remark this answers is not in the document]
Comment (NikhilBirgade) [reply btree/btree.c:3137]: Done.

File: btree/btree.c [hunk 16]
```diff
@@ -3108,6 +3108,145 @@ static void ff_rec_del_credit(const struct nd *node, m0_bcount_t ksize,
...
+ *		ii. We worry about this issue if we have a way to fix the tree
+ *		while online by having a redundant tree which can be used to fix
+ *		this broken tree. In which case we want to make sure that the
+ *		new record is not added now but only after the current tree is
+ *		fixed.
```
[the remark this answers is not in the document]
Comment (NikhilBirgade) [reply btree/btree.c:3142]: Done.

File: btree/btree.c [hunk 17]
```diff
@@ -3108,6 +3108,145 @@ static void ff_rec_del_credit(const struct nd *node, m0_bcount_t ksize,
...
+ *
+ * These are the proposal for supporting CRCs in the node:
```
[the remark this answers is not in the document]
Comment (NikhilBirgade) [reply btree/btree.c:3144]: Done.

File: btree/btree.c [hunk 18]
```diff
@@ -3108,6 +3108,145 @@ static void ff_rec_del_credit(const struct nd *node, m0_bcount_t ksize,
...
+ * two ways:
+ *	a. Motr will checksum all the keys and store checksum in the node header.
+ *		i. The checksum is verified when node is loaded from the disk.
+ *		ii. The checksum is updated when record is added/deleted.
+ *		iii. The checksum is not updated when record is updated since
+ *		the keys do not change in update operation.
+ *
+ *		Pros: Checksum will require less space.
+ *		Cons: On record addition/deletion, whole node needs to be
+ *		traversed for calculating CRC of the keys.
```
[the remark this answers is not in the document]
Comment (NikhilBirgade) [reply btree/btree.c:3160]: Done.

File: btree/btree.c [hunk 19]
```diff
@@ -3132,16 +3132,19 @@ static void ff_rec_del_credit(const struct nd *node, m0_bcount_t ksize,
...
- * These are the proposal for supporting CRCs in the node:
+ *		whatever location we reach and continue. This is because the
+ *		path to the correct location where the record should be inserted
+ *		is corrupted anyways and there is no way to recover the original
+ *		path leading to the correct location.
+ *		ii. We maintain a redundant tree and during traversal from the
+ *		root node towards the leaf node we confirm the CRC of every Key
+ *		which we encounter. If at any point we discover the Key node is
+ *		corrupted then we fix this corrupt Key using the copy from the
+ *		redundant tree and continue. This way the btree is corrupted
```
Comment (shashank-parulekar) [inline btree/btree.c:3143]: s/corrupted/corrected/

**Review by shashank-parulekar — APPROVED** [review]

File: btree/btree.c (hunk 19 above)
Comment (NikhilBirgade) [reply btree/btree.c:3143]: Done.

**Pull request merged**